### PR TITLE
 False positive when (one of) the callback function name(s) in #date_date_callbacks/#date_time_callbacks is a variable. 

### LINKED ISSUE #553: False positive when (one of) the callback function name(s) in `#date_date_callbacks`/`#date_time_callbacks` is a variable.
# Bug report
_(Yes, it's me again...)_

With #551 we introduced a false positive when (one of) the callback function name(s) in `#date_date_callbacks`/`#date_time_callbacks` is a variable.

### Code snippet that reproduces the problem
Code taken from `\Drupal\KernelTests\Core\Datetime\DatelistElementFormTest`

```
  public function datelistDateCallbackTrusted(array &$element, FormStateInterface $form_state, DrupalDateTime $date = NULL) {
    $element['datelistDateCallbackExecuted'] = [
      '#value' => TRUE,
    ];
    $form_state->set('datelistDateCallbackExecuted', TRUE);
  }

  public function buildForm(array $form, FormStateInterface $form_state, string $date_callback = 'datelistDateCallbackTrusted') {

    $form['datelist_element'] = [
      '#title' => 'datelist test',
      '#type' => 'datelist',
      '#default_value' => new DrupalDateTime('2000-01-01 00:00:00'),
      '#date_part_order' => [
        'month',
        'day',
        'year',
        'hour',
        'minute', 'ampm',
      ],
      '#date_text_parts' => ['year'],
      '#date_year_range' => '2010:2020',
      '#date_increment' => 15,
      '#date_date_callbacks' => [[$this, $date_callback]],
    ];

    $form['submit'] = [
      '#type' => 'submit',
      '#value' => t('Submit'),
    ];

    return $form;
  }
```

returns:

```
		-
			message: "#^\\#date_date_callbacks callback array\\{\\$this\\(Drupal\\\\KernelTests\\\\Core\\\\Datetime\\\\DatelistElementFormTest\\), string\\} at key '0' is not callable\\.$#"
			count: 1
			path: tests/Drupal/KernelTests/Core/Datetime/DatelistElementFormTest.php
```

### PR DESCRIPTION
**Pull request by Boegie**
Tries to resolve #553

### CODE REVIEW

File: src/Rules/Drupal/TrustedCallbackRule.php [hunk 1]
```diff
@@ -172,6 +174,13 @@ private function doProcessNode(Node\Expr $node, Scope $scope, string $keyChecked
 
         foreach ($type->getConstantArrays() as $constantArrayType) {
             if (!$constantArrayType->isCallable()->yes()) {
+                if ($constantArrayType->getItemType() instanceof UnionType
+                    && !($constantArrayType->getValuesArray()->getValueTypes()[1] instanceof ConstantStringType)) {
```
Comment (Boegie) [inline src/Rules/Drupal/TrustedCallbackRule.php:178]: This makes my eyes bleed, but couldn't find a decent way to determine if the right hand side of the UnionType is not a ConstantStringType, but a StringType, which (in my undestandi9ng of PHPStan magic) makes this a variable.

**Comment by Boegie** [comment]
Failing check is the new baseline, with correctly removed suppressions.

**Review by mglaman — CHANGES_REQUESTED** [review]
Renaming the class would be a breaking change as folks may have manually configured all of their rules (I don't know who would have, but I've been saving renaming/splitting this out for 1.2.x.)

**Comment by Boegie** [comment]
It seems one is never free of The Chains Of BC, even in here...

I'll revert the renaming.

**Comment by mglaman** [comment]
I opened https://github.com/mglaman/phpstan-drupal/issues/560 to take action on this, finally

File: src/Rules/Drupal/RenderCallbackRule.php [hunk 2]
```diff
@@ -172,6 +174,13 @@ private function doProcessNode(Node\Expr $node, Scope $scope, string $keyChecked
 
         foreach ($type->getConstantArrays() as $constantArrayType) {
             if (!$constantArrayType->isCallable()->yes()) {
+                if ($constantArrayType->getItemType() instanceof UnionType
+                    && !($constantArrayType->getValuesArray()->getValueTypes()[1] instanceof ConstantStringType)) {
```
Comment (Boegie) [inline src/Rules/Drupal/RenderCallbackRule.php:178]: This makes my eyes bleed, but couldn't find a decent way to determine if the right hand side of the UnionType is not a ConstantStringType, but a StringType, which (in my undestandi9ng of PHPStan magic) makes this a variable.

Comment (mglaman) [reply src/Rules/Drupal/RenderCallbackRule.php]: I think I found a way! Took a bit though.

**Review by mglaman — APPROVED** [review]

File: src/Rules/Drupal/RenderCallbackRule.php (hunk 2 above)
Comment (Boegie) [reply src/Rules/Drupal/RenderCallbackRule.php]: Nice, not as "grokable" as expected, but that's what's the PHPStan issue is for.

**Pull request merged**